### ISSUE TITLE
prefer-stateless-function triggers despite meeting all criteria with PureComponent

### ISSUE THREAD

**Issue by krainboltgreene** [yannickcr/eslint-plugin-react#968]
criteria:

When true the rule will ignore Components extending from React.PureComponent that use this.props or this.context.

  - [x] extends from `PureComponent`
  - [x] uses `this.props`

eslint-plugin-react version:

``` js
"version": "6.7.1"
```

eslint file:

``` js
module.exports = {
  parser: "babel-eslint",
  "parserOptions": {
    "ecmaFeatures": {
      "jsx": true
    }
  },
  plugins: [
    "babel",
    "react"
  ],
  env: {
    es6: true
  },
  rules: {
    "react/prefer-stateless-function": ["warn", {ignorePureComponents: true}]
}
```

Code:

``` jsx
import React, {PureComponent, PropTypes} from "react"
import {connect} from "react-redux"

import Layout from "../Layout"

export default connect()(class SignUp extends PureComponent {
  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    form: PropTypes.shape({
      email: PropTypes.string.isRequired,
      password: PropTypes.string.isRequired
    }).isRequired
  }
  render () {
    const {form} = this.props

    return <Layout subtitle="Join our website!">
      <section>
        <input type="email" value={form.email} />
        <input type="password" value={form.password} />
      </section>
    </Layout>
  }
})
```

Error:

<img width="1060" alt="screen shot 2016-11-20 at 9 20 17 pm" src="https://cloud.githubusercontent.com/assets/334809/20471499/27572fda-af67-11e6-9900-35138db9d0ff.png">

**Comment by ljharb** [comment]
If you use `React.PureComponent` instead of the named import, does the same thing happen?

**Comment by krainboltgreene** [comment]
Yes.
<img width="1021" alt="screen shot 2016-11-20 at 9 35 12 pm" src="https://cloud.githubusercontent.com/assets/334809/20471785/3eee0860-af69-11e6-9cb2-dba707bfecac.png">

**Comment by krainboltgreene** [comment]
If I remove the currying, it no longer warns me. It seems like it gets confused by the `connect()` currying.

**Comment by lucasbento** [comment]
Same thing happening here even with a simple component:

![image](https://user-images.githubusercontent.com/6207220/30416106-254d8456-992b-11e7-8af1-0172e49522ac.png)

```
// package.json
"eslint": "^4.6.1",
"eslint-plugin-react": "^7.3.0",
```

```js
// .eslintrc
{
  "extends": "airbnb",
  "parser": "babel-eslint",
  "rules": {
    "no-console": 0,
    "max-len": [
      1,
      120,
      2
    ],
    "no-param-reassign": [
      2,
      {
        "props": false
      }
    ],
    "no-continue": 0,
    "no-underscore-dangle": 0,
    "generator-star-spacing": 0,
    "class-methods-use-this": 0,
    "import/no-extraneous-dependencies": 0,
    "no-use-before-define": 0,
    "react/jsx-filename-extension": [1, { "extensions": [".js", ".jsx"] }],
    "react/prefer-stateless-function": [2, { "ignorePureComponents": true }]
  },
  "env": {
    "jest": true
  }
}
```

**Comment by jseminck** [comment]
Strange. This case is supported and has tests for it: https://github.com/yannickcr/eslint-plugin-react/blob/master/tests/lib/rules/prefer-stateless-function.js#L42

**Comment by ljharb** [comment]
@lucasbento are you setting your React version correctly in eslintrc? PureComponent doesn't exist before v15.2 or so

**Comment by Lepozepo** [comment]
Seeing the same issue having just created a new project with `create-react-native-app`

**Comment by MarkoCen** [comment]
I found some strange behavior with eslint 4.9.0 and react 16.0.0,:
```javascript
import React, { PureComponent } from 'react';

// this component would cause error 'react/prefer-stateless-function'
class ArticleCard extends PureComponent {
    render() {
        return (
            <div />
        );
    }
}

// if I add props reference in component render function, the error would disappear
class ArticleCard extends PureComponent {
    render() {
        const { id } = this.props;

        return (
            <div />
        );
    }
}
```
```json
// .eslintrc.json
{
  "env": {
    "browser": true,
    "commonjs": true,
    "es6": true
  },
  "extends": "airbnb",
  "rules": {
    "indent": ["error", 4],
    "linebreak-style": "off",
    "comma-dangle": "off",
    "arrow-body-style": "off",
    "no-trailing-spaces": ["error", {
      "skipBlankLines": true,
      "ignoreComments": true
    }],
    "function-paren-newline": "off",
    "object-curly-newline": ["error", {
      "multiline": true
    }],
    "global-require": "off",

    "react/jsx-indent": ["error", 4],
    "react/jsx-closing-bracket-location": ["error", "after-props"],
    "react/prefer-stateless-function": ["error", { "ignorePureComponents": true }],
    "jsx-a11y/anchor-is-valid": ["error", {
      "components": []
    }]
  }
}
```

```
// package.json
"react": "^16.0.0",
"react-dom": "^16.0.0",

"eslint": "4.9.0",
"eslint-config-airbnb": "^16.1.0",
"eslint-loader": "^1.9.0",
"eslint-plugin-import": "2.7.0",
"eslint-plugin-jsx-a11y": "6.0.2",
"eslint-plugin-react": "7.4.0",
```

**Comment by ljharb** [comment]
Given that we have passing tests, I'm going to close this. If you're still seeing issues in the latest version, please file a new issue.
